shell/platform/embedder/embedder.cc

```diff
@@ -712,6 +712,42 @@ inline flutter::PointerData::SignalKind ToPointerDataSignalKind(
   return flutter::PointerData::SignalKind::kNone;
 }
 
+// Returns the buttons for PointerData for the given buttons and change from a
+// FlutterPointerEvent.
+inline int64_t ToPointerDataButtons(int64_t buttons,
+                                    flutter::PointerData::Change change) {
+  switch (change) {
+    case flutter::PointerData::Change::kDown:
+    case flutter::PointerData::Change::kMove:
+    case flutter::PointerData::Change::kCancel:
+      // These kinds of change should always have a non-zero `buttons`.
+      // In case of violation, it synthesizes a primary button and log a warning
+      // to inform the embedder to send the correct buttons.
+      if (buttons == 0) {
+        // TODO(tongmu): Log a warning to inform the embedder to send the
+        // correct buttons. See
+        // https://github.com/flutter/flutter/issues/32052#issuecomment-489278965
+        return flutter::PointerButton::kPrimary;
+      }
+      return buttons;
+
+    case flutter::PointerData::Change::kAdd:
+    case flutter::PointerData::Change::kRemove:
+    case flutter::PointerData::Change::kHover:
+    case flutter::PointerData::Change::kUp:
+      // These kinds of change should always have a zero `buttons`. However we
+      // are not synthesizing a 0 until a solid reason is found.
```

Comment:
I think we should remove this comment; I don't think we should ever plan to fix up code where embedders are explicitly doing the wrong thing. Code to see buttons is all only here for backwards compatibility with embedders that never had to send buttons before, not to error-proof incorrect implementations of embedders.

Reply:
Got it.